File: xabber/src/main/java/com/xabber/android/data/extension/references/mutable/filesharing/FileSources.java
package com.xabber.android.data.extension.references.mutable.filesharing;

import org.jivesoftware.smack.util.XmlStringBuilder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class FileSources {

    public static final String SOURCES_ELEMENT = "sources";
    public static final String URI_ELEMENT = "uri";

    private List<String> uris;

    public FileSources() {}

    public FileSources(List<String> uris) {
        this.uris = uris;
    }

    public void addSource(String uri) {
        if (uris == null) uris = new ArrayList<String>();
        if (uri != null) uris.add(uri);
    }

    public void addSources(List<String> newUris) {
        if (uris == null) uris = new ArrayList<String>();
        uris.addAll(newUris);
    }

    public List<String> getSources() {
        return uris == null ? null : Collections.unmodifiableList(uris);
    }

    public CharSequence toXML() {
        XmlStringBuilder xml = new XmlStringBuilder();
        if (uris != null && !uris.isEmpty()) {
            xml.openElement(SOURCES_ELEMENT);
            for (String uri : uris) {
                xml.openElement(URI_ELEMENT);
                xml.append(uri);
                xml.closeElement(URI_ELEMENT);
            }
            xml.closeElement(SOURCES_ELEMENT);
        }
        return xml;
    }
}

File: xabber/src/main/java/com/xabber/android/data/extension/references/mutable/filesharing/FileSources.kt
package com.xabber.android.data.extension.references.mutable.filesharing

import org.jivesoftware.smack.util.XmlStringBuilder

/**
 * Represents a <sources> element at file-sharing reference according to [XEP-FILE]
 * The <sources/> element is used to define available sources where clients can download the file.
 * This element must contain at least one child element <uri/>.
 * The <uri/> element has information about uri.
 *
 * @link https://xabber.com/protocol/otb
 */
class FileSources(val uris: List<String>) {

    constructor(uri: String): this(listOf<String>(uri))

    fun toXML() = XmlStringBuilder().apply {
        openElement(SOURCES_ELEMENT)

        for (uri in uris){
            optElement(URI_ELEMENT, uri)
        }

        closeElement(SOURCES_ELEMENT)
    }

    companion object {
        const val SOURCES_ELEMENT = "sources"
        const val URI_ELEMENT = "uri"
    }

}